### DISCUSSION TITLE
Over-riding environment name in console?

### DISCUSSION THREAD

**Discussion by jrochkind** [rails/rails#52809] [Ideas]
After upgrading Rails 7.2, I noticed that the rails console prompt by default includes the Rails.environment name, with special red "danger" highlighting for production. 

This is nice!  See https://github.com/rails/rails/blob/d437ae311f1b9dc40b442e40eb602e020cec4e49/railties/lib/rails/commands/console/irb_console.rb#L122-L132

But I, like I think many people, use `config/production.rb` for both actual production and a staging environment. I think this is common?  It was less helpful to have both production and staging marked with a red `prod`. 

The best way I could figure out to fix this was with a prepend override, which also required some somewhat mysterious `requires`, that I'm not sure how stable they are (two are required, and they do not match the name of the class being overridden. 

```ruby
 require "rails/console/methods"
 require "rails/commands/console/irb_console"

# Over-ride rails method that puts the env in console prompt,
# so console prompt in staging says 'staging'.
#
# even though our staging uses "production" env, we want it to
# show up differently in console, so prompt can serve it's cautionary function to
# distinguish production

module OverrideIrbConsoleColorizedEnv
  def colorized_env
    if ScihistDigicoll::Env.staging?
      IRB::Color.colorize("staging", [:YELLOW])
    else
      super
    end
  end
end

Rails::Console::IRBConsole.prepend OverrideIrbConsoleColorizedEnv
```

Is there a better safer way to do this?  Would it be helpful for this to be configurable somehow more intentionally?

**Comment by skipkayhil** [comment]
You can customize the object used as the console with [`config.console`](https://guides.rubyonrails.org/configuring.html#config-console), so you could make a subclass of `IRBConsole`:

```
require ... # IRBConsole

class MyConsole < IRBConsole
  def colorized_env
end
```

and then in `config/application.rb` you should be able to

```
class Application < Rails::Application
  console do
    require ... # my_console
    config.console = MyConsole.new(Rails.application)
  end
end
```

**Reply by jrochkind** [reply]
Oh that is a good hint. Perhaps better than what I came up with, although effectively about the same! Thanks!

I guess this is considered acceptable as a way to customize for this use case?

**Comment by schmijos** [comment]
This is what I did. Poor man's figuring out where it runs:

```rb
console do
  require "rails/commands/console/irb_console.rb"

  class CloudyConsole < Rails::Console::IRBConsole
    def colorized_env
      if Rails.env.production?
        if ENV.fetch("HEROKU_APP_NAME", "")[/-develop|-staging/]
          return IRB::Color.colorize("prod", [:YELLOW])
        end

        if `hostname`[/develop-/]
          return IRB::Color.colorize("prod", [:YELLOW])
        end
      end

      super
    end
  end

  config.console = CloudyConsole.new(Rails.application)
end
```

(now seriously considering to use custom environments like `staging.rb` instead)
